smithy-model/src/main/java/software/amazon/smithy/model/loader/AbstractMutableModelFile.java

```diff
@@ -285,10 +260,106 @@ private <S extends Shape, B extends AbstractShapeBuilder<? extends B, S>> Option
                 builder.addTrait(trait);
             }
             return Optional.of(builder.build());
+        } catch (IllegalStateException e) {
+            if (builder.getShapeType() == ShapeType.MEMBER && ((MemberShape.Builder) builder).getTarget() == null) {
+                events.add(ValidationEvent.builder()
+                        .severity(Severity.ERROR)
+                        .id(Validator.MODEL_ERROR)
+                        .shapeId(builder.getId())
+                        .sourceLocation(builder.getSourceLocation())
+                        .message("Member target was elided, but no parent shape contained a matching member name or "
+                                + "resource identifier name.")
+                        .build());
+                return Optional.empty();
+            }
+            throw e;
         } catch (SourceException e) {
             events.add(ValidationEvent.fromSourceException(e, "", builder.getId()));
             resolvedTraits.clearTraitsForShape(builder.getId());
             return Optional.empty();
         }
     }
+
+    private static class ApplyMixin implements PendingShapeModifier {
```

Comment:
I really like this refactor :)

It might be nice to just move this to the upper level and keep it package-private to make the code easier to read.